### PR TITLE
topology translation of in-tree vSphere volume to vSphere CSI

### PR DESCRIPTION
**Pull request by divyenpatel**
#### What type of PR is this?
/kind bug

Optionally add one or more of the following kinds if applicable:
/kind regression

#### What this PR does / why we need it:
vSphere CSI driver does not understand legacy topology labels in-tree vSphere plugin is using.
This PR is adding topology translation from in-tree vSphere volume to vSphere CSI.

This PR is translating topology GA/topology beta labels to `topology.csi.vmware.com/region` and `topology.csi.vmware.com/zone`.

#### Special notes for your reviewer:
Pre-req for topology translation support.
vSphere CSI Driver must be installed using following vSphere Config Secret

```ini
[Labels]
region = k8s-region
zone = k8s-zone
[TopologyCategory "k8s-region"]
Label = "topology.csi.vmware.com/region"
[TopologyCategory "k8s-zone"]
Label = "topology.csi.vmware.com/zone"
```

Unit test execution

```sh
% go test -v -run  TestTranslatevSphereInTreeStorageClassToCSI
=== RUN   TestTranslatevSphereInTreeStorageClassToCSI
    vsphere_volume_test.go:113: Testing expect error when sc is nil
    vsphere_volume_test.go:113: Testing translate with no parameter
    vsphere_volume_test.go:113: Testing translate with unknown parameter
    vsphere_volume_test.go:113: Testing translate with storagepolicyname and datastore
    vsphere_volume_test.go:113: Testing translate with fstype
    vsphere_volume_test.go:113: Testing translate with storagepolicyname and fstype
    vsphere_volume_test.go:113: Testing translate with no parameter and allowedTopology
    vsphere_volume_test.go:113: Testing translate with storagepolicyname and allowedTopology
    vsphere_volume_test.go:113: Testing translate with storagepolicyname and allowedTopology beta labels
    vsphere_volume_test.go:113: Testing translate with raw vSAN policy parameters, datastore and diskformat
    vsphere_volume_test.go:113: Testing translate with all parameters
--- PASS: TestTranslatevSphereInTreeStorageClassToCSI (0.00s)
PASS
ok      k8s.io/csi-translation-lib/plugins      0.381s
```

```sh
 % go test -v -run  TestTranslateVSphereInTreePVToCSI
=== RUN   TestTranslateVSphereInTreePVToCSI
    vsphere_volume_test.go:464: Testing expect error when in-tree vsphere PV is nil
    vsphere_volume_test.go:464: Testing translate valid vSphere in-tree PV to vSphere CSI PV
    vsphere_volume_test.go:464: Testing translate valid vSphere in-tree PV with beta topology labels to vSphere CSI PV
    vsphere_volume_test.go:464: Testing translate valid vSphere in-tree PV with GA topology labels to vSphere CSI PV
--- PASS: TestTranslateVSphereInTreePVToCSI (0.00s)
PASS
ok      k8s.io/csi-translation-lib/plugins      0.178s
```

```sh
% go test -v -run  TestTranslateVSphereCSIPVToInTree
=== RUN   TestTranslateVSphereCSIPVToInTree
    vsphere_volume_test.go:276: Testing expect error when pv is nil
    vsphere_volume_test.go:276: Testing expect error when pv.Spec.CSI is nil
    vsphere_volume_test.go:276: Testing translate valid vSphere CSI PV to vSphere in-tree PV
    vsphere_volume_test.go:276: Testing translate valid vSphere CSI PV with topology Node Affinity rules to vSphere CSI PV with topology labels
--- PASS: TestTranslateVSphereCSIPVToInTree (0.00s)
PASS
ok      k8s.io/csi-translation-lib/plugins      0.365s
```

#### Does this PR introduce a user-facing change?

```release-note
topology translation of in-tree vSphere volume to vSphere CSI.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go [hunk 1]
```diff
@@ -206,3 +227,109 @@ func (t *vSphereCSITranslator) GetCSIPluginName() string {
...
+					Values: exp.Values,
+				}
+			} else if exp.Key == v1.LabelFailureDomainBetaRegion || exp.Key == v1.LabelTopologyRegion {
+				newExp = v1.TopologySelectorLabelRequirement{
+					Key:    regionkey,
+					Values: exp.Values,
+				}
+			} else {
+				// Other topologies are passed through unchanged.
+				newExp = exp
```
Comment (shalini-b) [inline staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go:328]: We did not support any other topologies in VCP right?

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go [hunk 2]
```diff
@@ -206,3 +227,109 @@ func (t *vSphereCSITranslator) GetCSIPluginName() string {
...
+	for _, term := range terms {
+		newTerm := v1.TopologySelectorTerm{}
+		for _, exp := range term.MatchLabelExpressions {
+			var newExp v1.TopologySelectorLabelRequirement
+			if exp.Key == v1.LabelFailureDomainBetaZone || exp.Key == v1.LabelTopologyZone {
+				newExp = v1.TopologySelectorLabelRequirement{
+					Key:    zonekey,
+					Values: exp.Values,
+				}
+			} else if exp.Key == v1.LabelFailureDomainBetaRegion || exp.Key == v1.LabelTopologyRegion {
```
Comment (shalini-b) [inline staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go:321]: SC Allowed topologies did not allow region labels right? Why is this needed?

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go [hunk 3]
```diff
@@ -48,6 +48,16 @@ func TestTranslatevSphereInTreeStorageClassToCSI(t *testing.T) {
 			Values: []string{"region-a"},
 		},
 	}}
+	expectedTopologySelectorTerm := v1.TopologySelectorTerm{[]v1.TopologySelectorLabelRequirement{
+		{
+			Key:    vSphereCSITopologyZoneKey,
+			Values: []string{"zone-a"},
+		},
+		{
+			Key:    vSphereCSITopologyRegionKey,
```
Comment (shalini-b) [inline staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go:57]: Region label wasn't allowed in SC allowed topologies right?

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go [hunk 4]
```diff
@@ -194,6 +204,84 @@ func TestTranslateVSphereCSIPVToInTree(t *testing.T) {
...
+				},
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						VsphereVolume: &v1.VsphereVirtualDiskVolumeSource{
+							VolumePath: "[vsanDatastore] 6785a85e-268e-6352-a2e8-02008b7afadd/kubernetes-dynamic-pvc-68734c9f-a679-42e6-a694-39632c51e31f.vmdk",
+							FSType:     "ext4",
+						},
+					},
+					AccessModes: []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce},
+					NodeAffinity: &v1.VolumeNodeAffinity{
```
Comment (shalini-b) [inline staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go:261]: The In-tree PV will have node affinity terms too?

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go [hunk 5]
```diff
@@ -264,6 +352,65 @@ func TestTranslateVSphereInTreePVToCSI(t *testing.T) {
 			},
 			expErr: false,
 		},
+		{
+			name: "translate valid vSphere in-tree PV with topology labels to vSphere CSI PV",
+			intreePV: &v1.PersistentVolume{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:   "pvc-d8b4475f-2c47-486e-9b57-43ae006f9b59",
+					Labels: map[string]string{"failure-domain.beta.kubernetes.io/zone": "z1", "failure-domain.beta.kubernetes.io/region": "r1"},
```
Comment (shalini-b) [inline staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go:360]: Should we have another test for In-tree PV with GA labels?

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go (hunk 2 above)
Comment (divyenpatel) [reply staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go:321]: Yes, and the vSphere in-tree plugin does not read region while creating volume.
https://github.com/kubernetes/kubernetes/blob/06e107081e4e0182efb9daa0daf7f2ae12272704/pkg/volume/vsphere_volume/vsphere_volume.go#L370-L376

I am removing the translation of region then.

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go (hunk 3 above)
Comment (divyenpatel) [reply staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go:57]: removing it

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go (hunk 4 above)
Comment (divyenpatel) [reply staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go:261]: I am not sure about it, but if at all we have node affinity rules present on the PV, we need to translate it so the schedular can use the node with topology keys applied by the vSphere CSI Driver.

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go (hunk 5 above)
Comment (divyenpatel) [reply staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume_test.go:360]: yes

File: staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go (hunk 1 above)
Comment (divyenpatel) [reply staging/src/k8s.io/csi-translation-lib/plugins/vsphere_volume.go]: yes used common `translateAllowedTopologies` function, all other plugin is using.

**Comment by divyenpatel** [comment]
@shalini-b I have addressed your review comments here - https://github.com/kubernetes/kubernetes/pull/108611/commits/7aa43796c2ebfd58a35a6f4a8cb925cbb8c29740
